pkg/operator/configobserver/config_observer_controller.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const operatorStatusTypeConfigObservationFailing = "ConfigObservationFailing"
+const configObservationErrorConditionReason = "ConfigObservationError"
+const configObserverWorkKey = "key"
+
+type OperatorClient interface {
+	GetOperatorState() (spec *operatorv1.OperatorSpec, status *operatorv1.OperatorStatus, resourceVersion string, err error)
+	UpdateOperatorSpec(string, *operatorv1.OperatorSpec) (spec *operatorv1.OperatorSpec, resourceVersion string, err error)
+	UpdateOperatorStatus(string, *operatorv1.OperatorStatus) (status *operatorv1.OperatorStatus, resourceVersion string, err error)
+}
+
+// Listers is an interface which will be passed to the config observer funcs.  It is expected to be hard-cast to the "correct" type
+type Listers interface {
+	PreRunHasSynced() []cache.InformerSynced
+}
+
+// observeConfigFunc observes configuration and returns the observedConfig. This function should not return an
+// observedConfig that would cause the service being managed by the operator to crash. For example, if a required
+// configuration key cannot be observed, consider reusing the configuration key's previous value. Errors that occur
+// while attempting to generate the observedConfig should be returned in the errs slice.
+type ObserveConfigFunc func(listers Listers, existingConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error)
+
+type ConfigObserver struct {
+	operatorClient OperatorClient
+
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+
+	rateLimiter flowcontrol.RateLimiter
+	// observers are called in an undefined order and their results are merged to
+	// determine the observed configuration.
+	observers []ObserveConfigFunc
+
+	// listers are used by config observers to retrieve necessary resources
+	listers Listers
+}
+
+func NewConfigObserver(
+	operatorClient OperatorClient,
+	listers Listers,
+	observers ...ObserveConfigFunc,
+) *ConfigObserver {
+	c := &ConfigObserver{
+		operatorClient: operatorClient,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConfigObserver"),
+
+		rateLimiter: flowcontrol.NewTokenBucketRateLimiter(0.05 /*3 per minute*/, 4),
+		observers:   observers,
+		listers:     listers,
+	}
+
+	return c
+}
+
+// sync reacts to a change in prereqs by finding information that is required to match another value in the cluster. This
+// must be information that is logically "owned" by another component.
+func (c ConfigObserver) sync() error {
+	originalSpec, originalStatus, resourceVersion, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	spec := originalSpec.DeepCopy()
+	// don't worry about errors
```

Comment:
lol ? :-)

Reply:
> lol ? :-)

we stomp over bad data.  Added explanation.